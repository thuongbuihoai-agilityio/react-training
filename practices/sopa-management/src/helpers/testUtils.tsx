import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import {
  QueryClient,
  QueryClientProvider
} from 'react-query';
import {
  BrowserRouter,
  Router
} from 'react-router-dom';

export const renderRouterTest = (component: JSX.Element) => {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {component}
    </Router>
  );
};

export const renderWithRouterAndQuery = (children: JSX.Element) => {
  const queryClient = new QueryClient();

  return render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </BrowserRouter>
  );
};
