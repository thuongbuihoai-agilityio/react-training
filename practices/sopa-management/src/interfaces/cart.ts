export const enum QuantityType {
  increment = 'increment',
  decrement = 'decrement'
}
