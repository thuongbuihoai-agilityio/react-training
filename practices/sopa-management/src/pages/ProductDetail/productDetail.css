.detail {
  padding: var(--px-extra-normal) var(--py-medium);
  margin: 0 auto;
  width: var(--w-extra-large);
  min-height: 100vh;
}

.detail-product {
  display: flex;
}

.detail-image {
  display: flex;
  align-items: center;
  min-width: var(--w-x-large);
  height: var(--h-extra-large);
  background-color: var(--bg-primary);
  justify-content: center;
}

.detail-info {
  margin-left: var(--m-large);
}

.detail-name {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: var(--m-large);
}

.detail-size {
  margin: var(--m-large) 0;
}

.detail-text-size {
  margin-bottom: var(--m-regular);
}

.detail-rating {
  display: flex;
  align-items: center;
  width: 192px;
  justify-content: space-between;
  margin-top: var(--m-normal);
}

.detail-btn {
  margin: 0 0 var(--m-extra-medium) 1px;
}

.detail-description {
  font-size: var(--text-extra-sm);
  line-height: 22px;
}
