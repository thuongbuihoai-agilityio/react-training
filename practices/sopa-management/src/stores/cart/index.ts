import { create } from 'zustand';

// Interfaces
import { Product } from '@interfaces/product';
import { QuantityType } from '@interfaces/cart';

type CartType = {
  cart: Product[];
  setCart: (value: Product[]) => void;
  addToCart: (product: Product, size: string) => void;
  updateQuantity: (id: string, quantityType: QuantityType) => void;
  deleteProductInCart: (id: string) => void;
};

export const useCartStore = create<CartType>()((set) => ({
  cart: [],
  setCart: (data: Product[]) => set(() => ({ cart: data })),
  addToCart: (product: Product, size: string) => {
    set((state) => {
      const currentCart = state.cart || [];
      const existingProduct = currentCart.find(
        (item: Product) => item.id === product.id && item.size === size
      );

      let updatedCart;
      if (existingProduct) {
        updatedCart = currentCart.map((item: Product) =>
          item.id === product.id && item.size === size
            ? { ...item, quantity: item.quantity + 1 }
            : item
        );
      } else {
        updatedCart = [...currentCart, { ...product, quantity: 1, size: size }];
      }

      set({ cart: updatedCart });
      return { cart: updatedCart };
    });
  },

  updateQuantity: (id: string, quantityType: QuantityType) => {
    set((state) => {
      const currentCart = state.cart || [];
      const updatedCart = currentCart.map((item: Product) => {
        const newQuantity =
          quantityType === QuantityType.increment
            ? item.quantity + 1
            : item.quantity - 1;
        return item.id === id ? { ...item, quantity: newQuantity } : item;
      });

      set({ cart: updatedCart });
      return { cart: updatedCart };
    });
  },

  deleteProductInCart: (id: string) =>
    set((state) => ({
      cart: state.cart.filter((product) => product.id !== id)
    }))
}));
