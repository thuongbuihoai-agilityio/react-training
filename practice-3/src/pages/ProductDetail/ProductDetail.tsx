import axios from "axios";
import React, { useCallback, useContext, useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import Price from "@/components/Price/Price";
import Button from "@/components/common/Button/Button";
import Title from "@/components/common/Title/Title";
import Text from "@/components/Text/Text";
import ModalUpdate from "@/components/Modal/ModalUpdate/ModalUpdate";
import { PRODUCT_CRUD } from "@/constants/url";
import { ProductContext } from "@/context/ProductContext";
import "./productDetail.css";

const ProductDetails: React.FC = () => {
  const { products } = useContext(ProductContext);
  const { id } = useParams();
  const [isReload, setIsReLoad] = useState(false);
  const dataEl = products?.find((item: { id: string }) => item.id === id);
  const [productDetailNew, setProductDetailNew] = useState(dataEl);
  const [openModalUpdate, setOpenModalUpdate] = useState(false);

  const toggleModalUpdate = useCallback(() => {
    setOpenModalUpdate(!openModalUpdate);
  }, [openModalUpdate]);

  async function fetchMyAPI() {
    const result = await axios
    .get(`${PRODUCT_CRUD}/${id}`)
    .then((res) => {
      setProductDetailNew(res.data);
    })
    .catch(function (error) {
      alert(error);
    });
  }

  useEffect(() => {
    if(id) {
      fetchMyAPI();
    }
  }, [isReload]);

  return (
    <>
      <div data-testid="product-detail-page" className="productDetails">
        <div className="productDetails__img--left">
          <img className="product__image" src={productDetailNew?.images[0]} />
        </div>
        <div className="productDetails__img--right">
          {productDetailNew?.images.map((img: string, index: number) => (
            <img key={index} src={img} alt="This is image" />
          ))}
        </div>
        <div className="productDetails__info">
          <div className="productDetail__update">
            <Title className="productDetail__title" text={productDetailNew?.name} />
            <Button
              onClick={toggleModalUpdate}
              className="btn btn__update"
              text={<i className="fa fa-pen"></i>}
            />
          </div>
          <Price className="productDetail__price" value={productDetailNew?.price} />
          <input
            className="productDetails__input"
            min={0}
            type="number"
            value={productDetailNew?.quantity}
          />
          <Text text={productDetailNew?.description} />
        </div>
      </div>
      {openModalUpdate && (
        <ModalUpdate
          product={productDetailNew}
          isReload={isReload}
          setIsReLoad={setIsReLoad}
          hideModalUpdate={toggleModalUpdate}
        />
      )}
    </>
  );
};

export default ProductDetails;
