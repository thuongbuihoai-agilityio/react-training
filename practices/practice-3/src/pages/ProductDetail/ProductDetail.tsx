import React, { useCallback, useContext, useState } from "react";
import { useParams } from "react-router-dom";
import Price from "@/components/Price/Price";
import Title from "@/components/common/Title/Title";
import Text from "@/components/Text/Text";
import ModalUpdate from "@/components/Modal/ModalUpdate/ModalUpdate";
import { DataContext } from "@/context/DataContext";
import { Product } from "@/types/product";
import toast from "react-hot-toast";
import "./productDetail.css";

const ProductDetails: React.FC = () => {
  // use useParams to get id
  const { id } = useParams();
  const { products, setProducts } = useContext(DataContext);
  const dataElement = products?.find((item) => item.id === id);
  const [openModalUpdate, setOpenModalUpdate] = useState(false);

  // update product detail
  const updateProductDetail = (response: Product) => {
    try {
      setProducts(response as any);
    } catch (error) {
      toast.error(error as string);
    }
  };

  // handle toggle modal update
  const toggleModalUpdate = useCallback(() => {
    setOpenModalUpdate(!openModalUpdate);
  }, [openModalUpdate]);

  return (
    <>
      <div data-testid="product-detail-page" className="productDetails">
        <div className="productDetails__img--left">
          <img className="product__image" src={dataElement?.images[0]} />
        </div>
        <div className="productDetails__img--right">
          {dataElement?.images.map((img: string, key: number) => (
            <img key={key} src={img} />
          ))}
        </div>
        <div className="productDetails__info">
          <div className="productDetail__update">
            <Title className="productDetail__title" text={dataElement?.name} />
            <button
              data-testid="open-modal-update"
              className="btn btn__update"
              onClick={toggleModalUpdate}
            >
              <i className="fa fa-pen"></i>
            </button>
          </div>
          <Price className="productDetail__price" value={dataElement?.price} />
          <input
            className="productDetails__input"
            min={0}
            type="number"
            defaultValue={dataElement?.quantity}
          />
          <Text text={dataElement?.description} />
        </div>
      </div>
      {openModalUpdate && (
        <ModalUpdate
          product={dataElement!}
          hideModalUpdate={toggleModalUpdate}
          updateProductDetail={updateProductDetail}
          deleteImage={() => {}}
        />
      )}
    </>
  );
};

export default ProductDetails;
