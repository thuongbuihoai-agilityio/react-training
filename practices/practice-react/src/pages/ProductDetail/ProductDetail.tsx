import React, { useContext } from "react";
import { DataContext } from "@context/DataContext";
import { useParams } from "react-router-dom";
import { Text, Button, Price, Counter, NavigationBar } from "@components/common/index";
import background from "@assets/images/backgrounds/gray.png";
import ProductList from "@components/ProductList/ProductList";
import "./productDetail.css";

const ProductDetail: React.FC = () => {
  const { id } = useParams();
  const { products } = useContext(DataContext);
  const dataElement = products?.find((item) => item.productId === id);
  const productMore = products.filter((product) =>
    dataElement?.categoryId.includes(product.categoryId)
  );

  return (
    <div data-testid="product-detail" className="product-detail">
      <NavigationBar mode="dark" />
      <figure>
        <img
          className="product-detail-image"
          src={dataElement?.images.src}
          alt={dataElement?.images.alt}
        />
        <img
          className="product-detail-background"
          src={background}
          alt="This is background"
        />
      </figure>
      <div className="product-detail-info">
        <Text text={dataElement?.productName} size="large" decoration="" />
        <p className="product-detail-price">Price</p>
        <Price
          value={dataElement?.originalPrice.value}
          type="original"
          currency={dataElement?.originalPrice.currency}
        />
        <p className="product-detail-description"></p>
        <p className="product-detail-quantity">Quantity</p>
        <div className="product-detail-counter">
          <Counter counter={dataElement?.productQuantity} />
        </div>
        <div className="product-detail-btn">
          <Button text="Add to cart" type="large" />
        </div>
      </div>
      <div className="product-detail-more">
        <div className="product-detail-title">
          <Text text="Products you may like" decoration="dark" size="" />
        </div>
        <div className="product-detail-product-more">
          <ProductList
            type="medium-box"
            content="quantity"
            visibleCounter={true}
            productList={productMore}
          />
        </div>
      </div>
    </div>
  );
};

export default ProductDetail;
