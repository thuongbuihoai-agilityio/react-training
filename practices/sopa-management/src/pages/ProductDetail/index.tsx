import { memo, useCallback, useState } from 'react';
import { useParams } from 'react-router-dom';
import toast from 'react-hot-toast';
import { shallow } from 'zustand/shallow';

// Components
import {
  Button,
  ButtonType,
  Dropdown,
  Loading,
  Price,
  PriceType,
  Rating,
  RatingStar,
  SizeType,
  Text
} from '@components/common';

// Hooks
import { useFetchProductDetail } from '@hooks/useQuery';
import {
  useMutationEditProductInCart,
  useMutationPostProductToCart
} from '@hooks/useMutate';

// Constants
import { SIZE } from '@constants/common';
import { CONFIRM_MESSAGE } from '@constants/validate';

// Interfaces
import { Product } from '@interfaces/product';

// Stores
import { useCartStore } from '@stores/cart';

// Styles
import './productDetail.css';

const ProductDetail = () => {
  // use useParams to get id
  const { id } = useParams();
  const [selectedValue, setSelectedValue] = useState('');

  // mutate
  const { mutate: postProduct } = useMutationPostProductToCart();
  const { mutate: putProduct } = useMutationEditProductInCart();

  const { data: product, isLoading } = useFetchProductDetail(id);

  const {
    name,
    image,
    color,
    price,
    description,
    size
  } = product;

  const [carts, addToCart] = useCartStore(
    (state) => [
      state.cart,
      state.addToCart
    ],
    shallow
  );

  const handleAddToCart = () => {
    const newData = {
      ...product,
      quantity: 1,
      size: selectedValue || size
    };

    const currentCart = carts || [];
    const existingProductIndex = currentCart.findIndex(
      (item: Product) => item.id === product.id
    );

    if (existingProductIndex !== -1) {
      putProduct(
        {
          ...currentCart[existingProductIndex],
          quantity: currentCart[existingProductIndex].quantity + 1
        },
        {
          onSuccess: () => {
            addToCart(product, selectedValue || size);
          }
        }
      );
    } else {
      postProduct(newData, {
        onSuccess: () => {
          addToCart(product, selectedValue || size);
        }
      });
    }

    toast.success(CONFIRM_MESSAGE.ADD_SUCCESS);
  };

  const handleSelect = useCallback(
    (value?: string) => {
      setSelectedValue(value as string);
    },
    [selectedValue]
  );

  return (
    <div data-testid='detail' className='detail'>
      {isLoading ? (
        <Loading />
      ) : (
        <>
          <div data-testid='product-detail' className='detail-product'>
            <div className='detail-image'>
              <img src={image.url} alt={image.alt} />
            </div>
            <div className='detail-info'>
              <div className='detail-name'>
                <div className='detail-description'>
                  <Text text={name} type={SizeType.regular} />
                  <RatingStar
                    value='4.7(3205)'
                    type={SizeType.normal}
                    className='detail-rating'
                    classNameStar='rating-star'
                  />
                </div>
                <Price value={price} type={PriceType.tertiary} />
              </div>
              <Text text={`Color: ${color}`} type={SizeType.extraRegular} />
              <div className='detail-size'>
                <Text
                  text='Size'
                  className='detail-text-size'
                  type={SizeType.extraRegular}
                />
                <Dropdown
                  data={SIZE}
                  value={selectedValue || size}
                  onSetValue={(value?: string) => handleSelect(value)}
                />
              </div>
              <Button
                ariaLabel='Add to bag'
                children={`Add to bag $${price}`}
                type={ButtonType.secondary}
                className='detail-btn'
                onClick={handleAddToCart}
              />
              <Text className='detail-description' text={description} />
            </div>
          </div>
          <Rating />
        </>
      )}
    </div>
  );
};

export default memo(ProductDetail);
