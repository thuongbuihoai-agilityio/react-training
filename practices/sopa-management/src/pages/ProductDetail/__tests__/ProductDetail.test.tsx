// Libs
import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import ProductDetail from '@pages/ProductDetail';

// Helpers
import { renderWithRouterAndQuery } from '@helpers/testUtils';

// Mocks
import {
  MOCK_PRODUCT,
  MOCK_PRODUCTS
} from '@mocks/product';

// Hooks
import {
  useFetchCartProduct,
  useFetchProductDetail
} from '@hooks/useQuery';
import { useCartStore } from '@stores/cart';

// Mock useParams
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ id: '1' })
}));

// Mock useFetchProductDetail
jest.mock('@hooks/useQuery', () => ({
  useFetchProductDetail: jest.fn(),
  useFetchCartProduct: jest.fn()
}));

describe('ProductDetail component', () => {
  beforeEach(() => {
    useCartStore.setState({
      cart: MOCK_PRODUCTS,
    });
  });

  test('should render Loading component', () => {
    (useFetchProductDetail as jest.Mock).mockReturnValue({
      data: MOCK_PRODUCT,
      isLoading: true
    });

    (useFetchCartProduct as jest.Mock).mockReturnValue({
      data: MOCK_PRODUCT
    });

    const { getByTestId } = renderWithRouterAndQuery(<ProductDetail />);
    expect(getByTestId('loading-page')).toBeInTheDocument();
  });

  test('should render ProductDetail component', () => {
    (useFetchProductDetail as jest.Mock).mockReturnValue({
      data: MOCK_PRODUCT,
      isLoading: false
    });

    const { getByTestId } = renderWithRouterAndQuery(<ProductDetail />);
    expect(getByTestId('product-detail')).toBeInTheDocument();
  });

  test('renders product details when not loading', () => {
    const { getByTestId } = renderWithRouterAndQuery(<ProductDetail />);

    expect(getByTestId('detail')).toBeInTheDocument();
    expect(screen.getByText(MOCK_PRODUCT.name)).toBeInTheDocument();
    expect(
      screen.getByText(`Color: ${MOCK_PRODUCT.color}`)
    ).toBeInTheDocument();
    expect(screen.getByText(`Size`)).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /Add to bag/i })
    ).toBeInTheDocument();
  });

  test('updates the selected value when a different value is selected', async () => {
    const { getByTestId } = renderWithRouterAndQuery(<ProductDetail />);
    const dropdown = getByTestId('dropdown');
    const detail = getByTestId('detail');
    await userEvent.click(dropdown);

    const option1 = screen.getByText('Small');
    await userEvent.click(option1);

    expect(detail).toHaveTextContent('Small');
  });

  test('matches snapshot', () => {
    const { container } = renderWithRouterAndQuery(<ProductDetail />);
    expect(container).toMatchSnapshot();
  });
});
