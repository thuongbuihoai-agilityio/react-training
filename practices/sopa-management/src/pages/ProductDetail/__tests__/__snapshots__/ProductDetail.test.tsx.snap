// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail component matches snapshot 1`] = `
<div>
  <div
    class="detail"
    data-testid="detail"
  >
    <div
      class="detail-product"
      data-testid="product-detail"
    >
      <div
        class="detail-image"
      >
        <img
          alt="This is black shoes"
          src="/images/products/black-shoes.png"
        />
      </div>
      <div
        class="detail-info"
      >
        <div
          class="detail-name"
        >
          <div
            class="detail-description"
          >
            <p
              class=" text-regular"
              data-testid="text"
            >
              Black shoes
            </p>
            <div
              class="detail-rating"
            >
              <div
                class="rating-star"
                data-testid="rating-star"
              >
                <div
                  class="undefined icon-star"
                  data-testid="icon"
                />
                <div
                  class="undefined icon-star"
                  data-testid="icon"
                />
                <div
                  class="undefined icon-star"
                  data-testid="icon"
                />
                <div
                  class="undefined icon-star"
                  data-testid="icon"
                />
                <div
                  class="undefined icon-star"
                  data-testid="icon"
                />
              </div>
              <p
                class=" text-normal"
                data-testid="text"
              >
                4.7(3205)
              </p>
            </div>
          </div>
          <p
            class=" price-tertiary"
            data-testid="price"
          >
            $
            87
          </p>
        </div>
        <p
          class=" text-extra-regular"
          data-testid="text"
        >
          Color: Black
        </p>
        <div
          class="detail-size"
        >
          <p
            class="detail-text-size text-extra-regular"
            data-testid="text"
          >
            Size
          </p>
          <div
            class=" custom-dropdown"
            data-testid="dropdown"
          >
            <p
              class="dropdown-label text-"
              data-testid="text"
            >
              Large
            </p>
          </div>
        </div>
        <button
          aria-label="Add to bag"
          class="detail-btn btn btn-secondary"
          data-testid="button"
        >
          Add to bag $87
        </button>
        <p
          class="detail-description text-"
          data-testid="text"
        >
          Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.
        </p>
      </div>
    </div>
    <div
      class="rating"
      data-testid="rating"
    >
      <p
        class=" text-extra-medium"
        data-testid="text"
      >
        Ratings and Reviews
      </p>
      <p
        class="rating-info text-regular"
        data-testid="text"
      >
        An overwhelming 97% of reviewers have endorsed this product, recommending it wholeheartedly to their friends and family.
      </p>
      <p
        class="rating-point text-large"
        data-testid="text"
      >
        4.7
      </p>
      <div
        class="rating-review"
      >
        <div
          class="rating-vote"
          data-testid="rating-star"
        >
          <div
            class="undefined icon-star"
            data-testid="icon"
          />
          <div
            class="undefined icon-star"
            data-testid="icon"
          />
          <div
            class="undefined icon-star"
            data-testid="icon"
          />
          <div
            class="undefined icon-star"
            data-testid="icon"
          />
          <div
            class="undefined icon-star"
            data-testid="icon"
          />
        </div>
        <p
          class=" text-normal"
          data-testid="text"
        >
          Based on 3,205 reviews
        </p>
      </div>
      <hr />
    </div>
  </div>
</div>
`;
